Content.Server/Discord/DiscordLink/DiscordLink.cs

```diff
@@ -0,0 +1,255 @@
+            _client = null;
+            return;
+        }
+
+        // If the Guild ID is empty OR the prefix is empty, we don't want to connect to Discord.
+        if (_configuration.GetCVar(CCVars.DiscordGuildId) == string.Empty || _configuration.GetCVar(CCVars.DiscordPrefix) == string.Empty)
+        {
+            // This is a warning, not info, because it's a configuration error.
+            // It is valid to not have a Discord token set which is why the above check is an info.
+            // But if you have a token set, you should also have a guild ID and prefix set.
+            _sawmill.Warning("No Discord guild ID or prefix specified, not connecting.");
+            _client = null;
+            return;
+        }
+
+        _botToken = token;
+        // Since you cannot change the token while the server is running / the DiscordLink is initialized,
+        // we can just set the token without updating it every time the cvar changes.
+
+        _client.Ready += () =>
+        {
+            _sawmill.Info("Discord client ready.");
+            return Task.CompletedTask;
+        };
+
+        Task.Run(async () =>
+        {
+            try
+            {
+                await LoginAsync(token);
+            }
+            catch (Exception e)
+            {
+                _sawmill.Error("Failed to connect to Discord!", e);
+            }
+        });
+    }
+
+    public void Shutdown()
+    {
+        if (_client != null)
+        {
+            _sawmill.Info("Disconnecting from Discord.");
+            _client.Log -= Log;
+
+            // Unsubscribe from the events.
+            _client.MessageReceived -= OnCommandReceivedInternal;
+            _client.MessageReceived -= OnMessageReceivedInternal;
+
+            _client.LogoutAsync();
+            _client.StopAsync();
+            _client.Dispose();
+            _client = null;
+        }
+
+        _configuration.UnsubValueChanged(CCVars.DiscordGuildId, OnGuildIdChanged);
+        _configuration.UnsubValueChanged(CCVars.DiscordPrefix, OnPrefixChanged);
+    }
+
+    public void PostInject()
+    {
+        _sawmill = _logManager.GetSawmill("discord.link");
+        _sawmillNet = _logManager.GetSawmill("discord.link.log");
+    }
+
+    private void OnGuildIdChanged(string guildId)
+    {
+        _guildId = ulong.TryParse(guildId, out var id) ? id : 0;
+    }
+
+    private void OnPrefixChanged(string prefix)
+    {
+        BotPrefix = prefix;
+    }
+
+    private async Task LoginAsync(string token)
+    {
+        DebugTools.Assert(_client != null);
+        DebugTools.Assert(_client.LoginState == LoginState.LoggedOut);
+
+        await _client.LoginAsync(TokenType.Bot, token);
+        await _client.StartAsync();
+
+
+        _sawmill.Info("Connected to Discord.");
+    }
+
+    private string FormatLog(LogMessage msg)
+    {
+        return msg.Exception is null ? $"{msg.Source}: {msg.Message}" : $"{msg.Source}: {msg.Exception}\n{msg.Message}";
```

Comment:
Should you not be printing the exception after the message in this case?

You'd get:
```
foobar: IndexOutOfRangeException
At stacktrace
At stacktrace
At stacktrace
At stacktrace
Error while doing X
```
which doesn't make much sense.